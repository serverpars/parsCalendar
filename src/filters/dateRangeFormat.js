/**
 * SPDX-FileCopyrightText: 2019 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
import moment from '@nextcloud/moment'
import { translate as t } from '@nextcloud/l10n'
import {
        formatJalaliDate,
        formatJalaliMonthYear,
        formatJalaliNumber,
        formatJalaliYear,
        getJalaliWeekInfo,
        isPersianLocale,
} from '@/utils/jalali.js'

/**
 * Formats a date-range depending on the user's current view
 *
 * @param {string | Date} value The date to format
 * @param {string} view The current view of the user
 * @param {string} locale Which locale to format it in
 * @return {string}
 */
export default (value, view, locale) => {
        if (isPersianLocale(locale)) {
                switch (view) {
                case 'timeGridDay':
                        return formatJalaliDate(value)

                case 'timeGridWeek': {
                        const { week, year } = getJalaliWeekInfo(value)

                        if (Number.isNaN(week) || Number.isNaN(year)) {
                                break
                        }

                        return t('calendar', 'Week {number} of {year}', {
                                number: formatJalaliNumber(week),
                                year: formatJalaliNumber(year),
                        })
                }

                case 'multiMonthYear':
                        return formatJalaliYear(value)

                case 'dayGridMonth':
                case 'listMonth':
                default:
                        return formatJalaliMonthYear(value)
                }
        }

        switch (view) {
        case 'timeGridDay':
                return moment(value).locale(locale).format('ll')

	case 'timeGridWeek':
		return t('calendar', 'Week {number} of {year}', {
			number: moment(value).locale(locale).week(),
			year: moment(value).locale(locale).weekYear(),
		})

	case 'multiMonthYear':
		return moment(value).locale(locale).format('YYYY')

	case 'dayGridMonth':
	case 'listMonth':
	default:
		return moment(value).locale(locale).format('MMMM YYYY')
	}
}
