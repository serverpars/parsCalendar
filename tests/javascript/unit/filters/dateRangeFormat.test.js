/**
 * SPDX-FileCopyrightText: 2019 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import dateRangeFormat from "../../../../src/filters/dateRangeFormat.js";
import { translate } from '@nextcloud/l10n'

jest.mock('@nextcloud/l10n')

describe('format/dateRangeFormat test suite', () => {

        beforeEach(() => {
                translate.mockClear()

                translate
                        .mockImplementation((app, str, params = {}) => str
                                .replace('{number}', params.number ?? '{number}')
                                .replace('{year}', params.year ?? '{year}'))
        })

        it('should provide a format for day view', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'timeGridDay', 'de')).toMatchSnapshot()
        })

        it('should provide a Jalali format for day view when locale is Persian', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'timeGridDay', 'fa')).toMatchSnapshot()
        })

        it('should provide a format for week view', () => {

                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'timeGridWeek', 'de')).toMatchSnapshot()
        })

        it('should provide a Jalali format for week view when locale is Persian', () => {

                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'timeGridWeek', 'fa')).toMatchSnapshot()
        })

        it('should provide a format for month view', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'dayGridMonth', 'de')).toMatchSnapshot()
        })

        it('should provide a Jalali format for month view when locale is Persian', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'dayGridMonth', 'fa')).toMatchSnapshot()
        })

        it('should provide month as fallback for unknown view', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'fooBarUnknownView', 'de')).toMatchSnapshot()
        })

        it('should provide Jalali month as fallback for unknown view when locale is Persian', () => {
                const date = new Date(2019, 0, 1, 0, 0, 0, 0)
                expect(dateRangeFormat(date, 'fooBarUnknownView', 'fa')).toMatchSnapshot()
        })

})
