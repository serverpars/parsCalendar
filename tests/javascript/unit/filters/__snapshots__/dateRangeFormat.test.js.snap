// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format/dateRangeFormat test suite should provide Jalali month as fallback for unknown view when locale is Persian 1`] = `"۱۳۹۷ دی"`;

exports[`format/dateRangeFormat test suite should provide a Jalali format for day view when locale is Persian 1`] = `"۱۱ دی ۱۳۹۷"`;

exports[`format/dateRangeFormat test suite should provide a Jalali format for month view when locale is Persian 1`] = `"۱۳۹۷ دی"`;

exports[`format/dateRangeFormat test suite should provide a Jalali format for week view when locale is Persian 1`] = `"Week ۴۱ of ۱٬۳۹۷"`;

exports[`format/dateRangeFormat test suite should provide a format for day view 1`] = `"1. Jan. 2019"`;

exports[`format/dateRangeFormat test suite should provide a format for month view 1`] = `"Januar 2019"`;

exports[`format/dateRangeFormat test suite should provide a format for week view 1`] = `"Week 1 of 2019"`;

exports[`format/dateRangeFormat test suite should provide month as fallback for unknown view 1`] = `"Januar 2019"`;
